Update CNI plugin to newest version; support ConfigLists

squeed opened the pull request:
**What this PR does / why we need it**: Updates the CNI network plugin to use the newest version of CNI. This brings with it plugin chaining and support for multiple versions.

**Special notes for your reviewer**: This libcni change is backwards-compatible - older plugins will work without any changes needed.

**Release note**:
```release-note
This adds support for CNI ConfigLists, which permit plugin chaining.
```

squeed commented:
@freehan

feiskyer commented:
@squeed Cool. When will CNI v0.5.0 be released?

squeed commented:
We were hoping to cut a -rc release today or tomorrow, and a full release a few days after that.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42202)
<!-- Reviewable:end -->

freehan commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -314,10 +315,15 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
 		return err
 	}
+	// Coerce the CNI result version
+	res, err := cnitypes020.GetResult(resT)
```
I assume this is backward compatible, right?

squeed replied:
Correct - this converts the result in to the 0.2.0 format, which is what kubenet already expects.

freehan commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -292,5 +318,23 @@ func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubec
+			"host_ip":        p.HostIP,
+		})
+	}
+	// TODO(cdc) CNI #373 needs to be merged first
```
I assume there will be follow up on this

squeed replied:
Yup :-). It was merged a few hours ago.

freehan commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -98,17 +99,40 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
+				glog.Warningf("Error converting CNI config file %s to list: %v", confFile, err)
+				continue
+			}
+
```
nit: remove this empty line

freehan commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -98,17 +99,40 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
+				continue
+			}
+		} else {
+
```
nit: remove this empty line

freehan reviewed (commented):
Look good overall. Waiting for CNI release. Will take a final look later

freehan commented:
@k8s-bot ok to test

freehan commented:
Need to run `hack/update-bazel.sh`

squeed commented:
0.5.0-rc1 was just released. Going to update the PR with that.

squeed commented:
Ok, all checks are passing. All that's left now is to wait for 0.5.0 to be released - should be sometime this week.

squeed commented:
CNI v0.5.0 has been updated. This can be considered for merge.

squeed commented:
Test failures seem like flakes...

squeed commented:
@freehan ready for review.

caseydavenport commented:
@squeed @freehan hmm, what's the status of / plan for this guy? Did this not make it into v1.6?

CC @thockin

thockin commented:
@freehan has been unplanned-out for a couple weeks, and I didn't have
bandwidth to pick up all his PRs.  This feels risky at this point, doesn't
it?  Is it the best fix?  Plugins still have to be updated, right?

On Mon, Mar 20, 2017 at 4:09 PM, Casey Davenport <notifications@github.com>
wrote:

> @squeed <https://github.com/squeed> @freehan <https://github.com/freehan>
> hmm, what's the status of / plan for this guy? Did this not make it into
> v1.6?
>
> CC @thockin <https://github.com/thockin>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/42202#issuecomment-287926376>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGkTmaN9yFvi8j8WzK10VAXbIs-eks5rnwcZgaJpZM4MNzCd>
> .
>

tomdee commented:
@thockin I think without this PR the hostport plugin can't be used.

thockin commented:
That's fine - kube has its own support for 1.6, plan was to try to
transition in 1.7

On Mon, Mar 20, 2017 at 4:19 PM, Tom Denham <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> I think without this PR the
> hostport plugin can't be used.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/42202#issuecomment-287928185>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVHk1Uno2ogR_Y99UNuMRiaigGUkcks5rnwlpgaJpZM4MNzCd>
> .
>

squeed commented:
It would be nice to get this in for 1.6.1, if possible. It does **not** require any plugin changes.

squeed commented:
@thockin @freehan care to review this? It's ready to go.

freehan commented:
I am back. Will review this. Thanks!

squeed commented:
There is a minor bug in cni v0.5.0 - there is a v0.5.2 release coming in a day or so. I will update the PR for that - so DO NOT MERGE for now please.

Commit:
deps: update containernetworking/cni to v0.5.2

Commit:
kubelet/networking: add support for cni ConfigLists, pass hostport parameters

** reason for this change **
CNI has recently introduced a new configuration list feature. This
allows for plugin chaining. It also supports varied plugin versions.

squeed commented:
@freehan 0.5.2 has been released and merged in to this PR - ready for review + merge.

freehan reviewed (commented):
Looks good overall. Just some nits.

freehan commented:
FYI @dnardo

squeed commented on Godeps/Godeps.json:
```diff
@@ -416,18 +416,33 @@
+			"Rev": "137b4975ecab6e1f0c24c1e3c228a50a3cfba75e"
+		},
+		{
+			"ImportPath": "github.com/containernetworking/cni/pkg/types/020",
```
Nope. Godeps is special :-)

squeed commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -294,5 +326,24 @@ func buildCNIRuntimeConf(podName string, podNs string, podInfraContainerID kubec
+			HostIP:        p.HostIP,
+		})
+	}
+	rt.CapabilityArgs = map[string]interface{}{
```
That's correct!

feiskyer replied:
It seems there is no official plugins (e.g. bridge) support port-mappings yet, is it right?

dcbw replied:
@feiskyer we don't intend to put that capability into the main plugins.  Instead, you would create a CNI ConfigList that includes the CNI "portmap" plugin here https://github.com/containernetworking/plugins/pull/1.  When the Kube CNI driver calls AddNetworkList(), libcni will execute each plugin in the list sequentially, passing the results from each plugin along to the next.  The "portmap" plugin consumes that result and sets up the mapings as required.

squeed commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -243,16 +275,16 @@ func (plugin *cniNetworkPlugin) GetPodNetworkStatus(namespace string, name strin
 	netconf, cninet := network.NetworkConfig, network.CNIConfig
-	glog.V(4).Infof("About to run with conf.Network.Type=%v", netconf.Network.Type)
-	res, err := cninet.AddNetwork(netconf, rt)
+	glog.V(4).Infof("About to run with conf.Network.Type=%v", netconf.Plugins[0].Network.Type)
```
Good point (though I kept it as original). I'll log both

squeed commented:
@freehan updated & rebased, thanks

freehan commented:
/lgtm
/approve

Thanks!

freehan commented:
Could you approve this? It needs owners at root.  @thockin

squeed commented:
Had to rebase. @freehan @thockin is there any chance of getting this reviewed and merged quickly?

feiskyer commented on pkg/kubelet/network/cni/cni.go:
```diff
@@ -96,17 +106,37 @@ func getDefaultCNINetwork(pluginDir, binDir, vendorCNIDirPrefix string) (*cniNet
+
 		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
-		vendorDir := vendorCNIDir(vendorCNIDirPrefix, conf.Network.Type)
+		vendorDir := vendorCNIDir(vendorCNIDirPrefix, confType)
```
Should we search vendor CNI dir for all plugin types, e.g. networks may have different type in `confList.Plugins`

squeed replied:
Maybe, but that is changing behavior. The aim of this PR is only to enable config lists.

squeed replied:
There is a lot of cleaning up that can be done with how k8s uses CNI but this PR is not the place for it, IMHO. Just trying to get the version bump out now.

freehan commented:
/lgtm

freehan commented:
@thockin for approve

freehan commented:
fyi @dnardo

squeed commented:
Fingers crossed that this wins the rebase race.

dcbw commented:
LGTM

thockin commented:
/approve

redbaron commented:
are there any plans to have it in 1.6 branch?

bboreham commented:
Can I request that the top description of this PR be extended to mention the portmap capability that it adds?

Really I would like it in the 1.7.0 release notes too, but I'm not sure if those can be changed retrospectively.
